### PR TITLE
Add `EDFRecordingExtractor` 

### PR DESCRIPTION
**Pull request by h-mayorquin**
I am adding a recording extractor that reads the European Data Format:
https://www.edfplus.info/

This is based on neo so is a neo extractor:

https://github.com/NeuralEnsemble/python-neo/blob/f62bd73bff79d84d3f4cca653b37c86064205a4e/neo/rawio/edfrawio.py

I also added tests trying to follow the format set up in the `test_neoextractors` file. 

This is the first time that I add one extractor so let me know what else is missing or could be improved.

Related to:
https://github.com/catalystneuro/neuroconv/issues/44

### CODE REVIEW

**Comment by h-mayorquin** [comment]
OK, so there is an error in the testing generated by the following situation:

The common neo test that `spikeinterface` uses for testing the opening of the files (`test_open`) loads the extractor almost at the beginning:
https://github.com/SpikeInterface/spikeinterface/blob/de8e6932eade03623b0a92f5b3ccada6d6e0cb7f/spikeinterface/extractors/tests/common_tests.py#L35

But then, to test the annotation propagation mechanism that I think @samuelgarcia added recently the extractor is loaded again:
https://github.com/SpikeInterface/spikeinterface/blob/de8e6932eade03623b0a92f5b3ccada6d6e0cb7f/spikeinterface/extractors/tests/common_tests.py#L69-L71

For the `neo.rawio.EDFRawIO` trying to open the file again results in the IO error shown in the full tests here. 

What should we do? Is this a problem with `EDFRawIO` or should the test not try to open the file again? 
What do you think @samuelgarcia @JuliaSprenger

**Comment by JuliaSprenger** [comment]
With the pyedf library in the background the file can be only opened once. In [the test] (https://github.com/SpikeInterface/spikeinterface/blob/de8e6932eade03623b0a92f5b3ccada6d6e0cb7f/spikeinterface/extractors/tests/common_tests.py#L69-L71) you could check if the IO has a `close` attribute and close the first instance before opening a second one.

**Comment by h-mayorquin** [comment]
> With the pyedf library in the background the file can be only opened once. In [the test] (
> 
> https://github.com/SpikeInterface/spikeinterface/blob/de8e6932eade03623b0a92f5b3ccada6d6e0cb7f/spikeinterface/extractors/tests/common_tests.py#L69-L71
> 
> ) you could check if the IO has a `close` attribute and close the first instance before opening a second one.

OK, thanks for your input. Let's say what @samuelgarcia thinks as this might present itself in the future once again with a different extractor.

One possible solution is to test the annotation in a separate test.

**Comment by samuelgarcia** [comment]
Hi Ramon,
this looks good.
yes, the double opening cause problem in edf.

You are right.
You can add a new method in RecordingCommonTestSuite line 68.
It should avoid the problem.

Lets got for this.

File: spikeinterface/extractors/tests/common_tests.py [hunk 1]
```diff
@@ -66,10 +66,17 @@ def test_open(self):
...
+                            
+    def test_neo_annotations(self):
+        for entity in self.entities:
 
+            if isinstance(entity, tuple):
+                path, kwargs = entity
+            elif isinstance(entity, str):
+                path = entity
+                kwargs = {}
+            if hasattr(self.ExtractorClass , "NeoRawIOClass"):
```
Comment (h-mayorquin) [inline spikeinterface/extractors/tests/common_tests.py:78]: I don't know if there is a better way of doing this. My other option was (example):
`self.ExtractorClass.__base__.__base__ == _NeoBaseExtractor`

File: spikeinterface/extractors/neoextractors/edf.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    all_annotations: bool, optional, default: False
+        Load exhaustively all annotations from neo.
+    """
+    mode = 'file'
+    NeoRawIOClass = 'EDFRawIO'
+
+    def __init__(self, file_path, stream_id=None, all_annotations=False):
+        neo_kwargs = {'filename': str(file_path)}
+        NeoBaseRecordingExtractor.__init__(self, stream_id=stream_id, all_annotations=all_annotations, **neo_kwargs)
+        self._kwargs.update({'file_path': str(file_path)})
```
Comment (bendichter) [inline spikeinterface/extractors/neoextractors/edf.py:27]:
```suggestion
        self._kwargs.update(file_path=str(file_path))
```

Comment (h-mayorquin) [reply spikeinterface/extractors/neoextractors/edf.py:27]: I am following the template that they have for all the other neo extractors on this one.

**Review by alejoe91 — APPROVED** [review]

**Pull request merged**